hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -439,6 +441,121 @@ struct Visitor : hilti::visitor::PreOrder<bool, Visitor> {
     }
 };
 
+struct TypeVisitor : visitor::PreOrder<bool, TypeVisitor> {
+    Stage _stage = Stage::COLLECT;
+    std::map<ID, bool> _used;
+
+    void collect(Node& node) {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    bool prune_uses(Node& node) { return false; }
+    bool prune_decls(Node& node) {
+        _stage = Stage::PRUNE_DECLS;
+
+        bool any_modification = false;
+
+        for ( auto i : this->walk(&node) )
+            if ( auto x = dispatch(i) )
+                any_modification = any_modification || *x;
+
+        return any_modification;
+    }
+
+    result_t operator()(const declaration::Type& x, position_t p) {
+        const auto type_id = x.typeID();
+
+        if ( ! type_id )
+            return false;
+
+        switch ( _stage ) {
+            case Stage::COLLECT:
+                // Record the type if not already known. If the type is part of an external API record it as used.
+                _used.insert({*type_id, x.linkage() == declaration::Linkage::Public});
+                break;
+
+            case Stage::PRUNE_USES: break;
+            case Stage::PRUNE_DECLS:
+                if ( ! _used.at(*type_id) ) {
+                    HILTI_DEBUG(logging::debug::GlobalOptimizer, util::fmt("removing unused type '%s'", *type_id));
```

Comment:
Something is flaky here with the test `spicy.optimizations.unused-types`. I sometimes see `Priv2` and `Priv3` not being optimized out.

Reply:
We should `return true` from this conditional, otherwise the edits will not converge to the same final state.